File: SudokuSolver.java
public class SudokuSolver {
    public static void main(String[] args) {

        //Input sudoku grid
        int[][] grid = {{0, 2, 0, 0, 0, 5, 0, 0, 0},
                        {0, 1, 5, 0, 0, 0, 0, 0, 0},
                        {0, 0, 0, 0, 0, 8, 7, 0, 3},
                        {0, 5, 1, 0, 0, 0, 0, 0, 0},
                        {0, 0, 9, 7, 0, 0, 0, 1, 0},
                        {0, 0, 0, 3, 0, 0, 0, 4, 6},
                        {0, 0, 0, 0, 8, 0, 0, 0, 1},
                        {7, 0, 0, 9, 3, 0, 0, 6, 0},
                        {0, 0, 0, 0, 0, 0, 4, 0, 8}};

        if(solveSudoku(grid)) {
            printSudoku(grid);
        }
        else {
            System.out.println("No solution");
        }
    }

    //Main method that solves the sudoku
    private static boolean solveSudoku(int[][] arr) {
        int[] newLoc = findEmptyLocation(arr);

        //If null, then sudoku has been solved
        if (newLoc == null) {
            return true;
        }

        int row = newLoc[0];
        int col = newLoc[1];

        for (int x = 1; x < 10; x++) {
            if (isValid(arr, row, col, x)) {
                arr[row][col] = x;

                if (solveSudoku(arr)) {
                    return true;
                }
                arr[row][col] = 0;
            }
        }
        return false;
    }

    private static void printSudoku(int[][] arr) {
        for (int i = 0; i < 9; i++) {
            StringBuilder row = new StringBuilder();
            for (int j = 0; j < 9; j++) {
                // Appends each element to the row
                row.append(arr[i][j]).append(" ");
            }
            System.out.println(row);
        }
    }

    private static boolean isValid(int[][] arr, int row, int col, int x) {

        // Checks if the number x is unique in the row
        for (int i = 0; i < 9; i++) {
            if (arr[row][i] == x) {return false;}
        }

        // Checks if the number x is unique in the column
        for (int i = 0; i < 9; i++) {
            if (arr[i][col] == x) {return false;}
        }

        // Checks if the number x is unique in it's square
        for (int i = 0; i < 3; i++){
            for (int j = 0; j < 3; j++) {
                if (arr[i + row - row%3][j + col - col%3] == x) {
                    return false;
                }
            }
        }
        // Returns true if x is unique in row, column, and the square
        return true;
    }

    // Method finds the next location in the grid that doesn't have a number
    private static int[] findEmptyLocation(int[][] arr) {
        for (int i = 0; i < 9; i++){
            for (int j = 0; j < 9; j++) {
                if (arr[i][j] == 0) {
                    return new int[]{i, j};
                }
            }
        }
        // Returns null if there is no more empty locations
        return null;
    }
}

File: sudokusolver.c
#include <stdio.h>

int solve_sudoku(int arr[9][9]);
int is_valid(int arr[9][9], int row, int col, int x);
void find_empty_location(int arr[9][9], int out[2]);
void print_sudoku(int arr[9][9]);

int main(void) {

    // Input unsolved sudoku grid
    int grid[9][9] = {{0, 2, 0, 0, 0, 5, 0, 0, 0},
                      {0, 1, 5, 0, 0, 0, 0, 0, 0},
                      {0, 0, 0, 0, 0, 8, 7, 0, 3},
                      {0, 5, 1, 0, 0, 0, 0, 0, 0},
                      {0, 0, 9, 7, 0, 0, 0, 1, 0},
                      {0, 0, 0, 3, 0, 0, 0, 4, 6},
                      {0, 0, 0, 0, 8, 0, 0, 0, 1},
                      {7, 0, 0, 9, 3, 0, 0, 6, 0},
                      {0, 0, 0, 0, 0, 0, 4, 0, 8}};

    if (solve_sudoku(grid)) 
        print_sudoku(grid);  
    
    else 
        printf("No solution found.\n");

    return 0;
}

// Main function that solves the sudoku using backtracking
int solve_sudoku(int arr[9][9]) {

    int next_loc[2] = {-1, -1}; // initalizes the next location to an invalid location
    find_empty_location(arr, next_loc);
    if (next_loc[0] == -1) { // if there is no next location, then the algorithm is done
        return 1;
    }

    int row = next_loc[0];
    int col = next_loc[1];

    // loops through the digits 1-9 for a given location
    for (int x = 1; x < 10; x++) {
        if(is_valid(arr, row, col, x) == 1) { // checks if the digit is valid at this row,column
            arr[row][col] = x;

            // Recursivly calls solver until no digit works in a location, then causes backtrack
            if (solve_sudoku(arr) == 1) {
                return 1;
            }
            // if run fails, reset to 0 then will backtrack
            arr[row][col] = 0;
        }
    }
    //triggers backtracking
    return 0;
}

// Checks if X is allowed to be placed in the grid at row,col
int is_valid(int arr[9][9], int row, int col, int x) {
    // Checks if the number x is unique in the row
    for (int i = 0; i < 9; i++) {
        if (arr[row][i] == x) {
            return 0;
        }
    }

    // Checks if the number x is unique in the column
    for (int j = 0; j < 9; j++) {
        if (arr[j][col] == x) {
            return 0;
        }
    }

    // Checks if the number x is unique in it's square
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            if (arr[i + row - row%3][j + col - col%3] == x) {
                return 0;
            }
        }
    }
    // Returns 1 if x is unique in row, column, and the square
    return 1;
}

// Finds the next location in the grid that doesn't have a number (0)
void find_empty_location(int arr[9][9], int out[2]) {
    for (int i = 0; i < 9; i++){
        for (int j = 0; j < 9; j++) {
            if (arr[i][j] == 0) {
                out[0] = i;
                out[1] = j;
                return;
            }
        }
    }
}

// Prints the supplied sudoku grid to the console with spaces between the numbers
void print_sudoku(int arr[9][9]) {
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            printf("%d ", arr[i][j]);
        }
        printf("\n");
    }
}
